README.md
# dhooks-lite

![version](https://img.shields.io/pypi/v/dhooks-lite)
![python](https://img.shields.io/pypi/pyversions/dhooks-lite)
![license](https://img.shields.io/github/license/ErikKalkoken/dhooks-lite)
![build](https://api.travis-ci.org/ErikKalkoken/dhooks-lite.svg?branch=master)
[![codecov](https://codecov.io/gh/ErikKalkoken/dhooks-lite/branch/master/graph/badge.svg)](https://codecov.io/gh/ErikKalkoken/dhooks-lite)
[![Documentation Status](https://readthedocs.org/projects/dhooks-lite/badge/?version=latest)](https://dhooks-lite.readthedocs.io/en/latest/?badge=latest)
[![Downloads](https://pepy.tech/badge/dhooks-lite)](https://pepy.tech/project/dhooks-lite)

>**Note**:<br>This repository has moved to a new locaton: https://gitlab.com/ErikKalkoken/dhooks-lite. This version has been archived.

## Contents

- [Overview](#Overview)
- [Functionality](#functionality)
- [Examples](#examples)
- [Installation](#installation)
- [Documentation](#documentation)
- [Contribution](#contribution)
- [Change Log](CHANGELOG.md)

## Overview

**dhooks-lite** is a library with a set of classes for interacting with Discord webhooks written in Python 3.

This library aims to differentiate itself from similar libraries with the following properties:

- is fully tested
- simple to use (only one way of doing things, same name of attributes and objects as in the [official Discord documentation](https://discordapp.com/developers/docs/resources/webhook#execute-webhook))
- has logging
- requests are automatically retried and have sensible timeouts

## Functionality

This library provides following functionality:

- Posting messages in Discord channels via webhooks (synchronous calls only)
- Attaching Embeds to messages
- Retrieve send reports and from Discord
- Retrieve HTTP status and headers from Discord, e.g. for implementing rate limit handling

## Examples

Here are some examples on how to use dhooks-lite in your Python scripts.

Note that you also find the source code of all examples in the `/examples` folder of this repo.

### Hello World

Minimal example for posting a message.

```python
from dhooks_lite import Webhook

hook = Webhook(DISCORD_WEBHOOK_URL)
hook.execute('Hello, World!')
```

![example1](https://i.imgur.com/t3mxMAJ.png)

### Posting with custom avatar

In this example we are setting username and avatar.

```python
from dhooks_lite import Webhook

hook = Webhook(
    DISCORD_WEBHOOK_URL,
    username='Bruce Wayne',
    avatar_url='https://i.imgur.com/thK8erv.png'
)
hook.execute('I am Batman!')
```

![example2](https://i.imgur.com/mseg2Yx.png)

### Complete example with embeds

Finally, here is an example for posting a message with two embeds and using all available features (shortened):

```python
import datetime
from dhooks_lite import Webhook, Embed, Footer, Image, Thumbnail, Author, Field

hook = Webhook(DISCORD_WEBHOOK_URL)
e1 = Embed(
    description='Only a few years ago, scientists stumbled upon an electrical current of cosmic proportions.(...)',
    title='Universe\'s highest electric current found',
    url='https://www.newscientist.com/article/mg21028174-900-universes-highest-electric-current-found/',
    timestamp=datetime.datetime.utcnow(),
    color=0x5CDBF0,
    footer=Footer(
        'Science Department',
        'https://i.imgur.com/Bgsv04h.png'
    ),
    image=Image('https://i.imgur.com/eis1Y0P.jpg'),
    thumbnail=Thumbnail('https://i.imgur.com/2A4k28x.jpg'),
    author=Author(
        'John Scientist',
        url='https://en.wikipedia.org/wiki/Albert_Einstein',
        icon_url='https://i.imgur.com/1JoHDw1.png'
    ),
    fields=[
        Field('1st Measurement', 'Failed'),
        Field('2nd Measurement', 'Succeeded')
    ]
)
e2 = Embed(description="TOP SECRET - Do not distribute!")

hook.execute(
    'Checkout this new report from the science department:',
    username='Bruce Wayne',
    avatar_url='https://i.imgur.com/thK8erv.png',
    embeds=[e1, e2],
    wait_for_response=True
)
```

![example2](https://i.imgur.com/RoWBh2n.png)

## Installation

You can install this library directly from PyPI:

```bash
pip install dhooks-lite
```

## Documentation

For a full documentation of all classes please see the official docs [here](https://dhooks-lite.readthedocs.io/en/latest/).

## Contribution

We welcome any contribution!

If you found a bug or have a feature request please raise an issue.

If you want to help further improve this library please feel free to issue a merge request. Please adhere to the following requirements in your change:

- Code should be compliant with [PEP8](https://www.python.org/dev/peps/pep-0008/)
- Full overage by unit tests
- All classes should be immutable
- All classes and their public methods must have docstring documentation
- All changes must be documented in the Change Log


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] - yyyy-mm-dd

## [1.0.0] - 2023-04-18

### Added

- Support for Python 3.9 & 3.10
- Improved type hints

## [0.6.1] - 2021-01-03

### Fixed

- Install aborts without requests: `ModuleNotFoundError: No module named 'requests'`

## [0.6.0] - 2021-01-02

### Added

- Ability to set an user agent for all requests
- Ability to set max retries to any value and disable it

### Changed

- Remove support for Python 3.5

## [0.5.1] - 2020-12-30

### Fixed

- Did not return a WebhookResponse object for http errors

## [0.5.0] - 2020-07-03

### Added

- Added new ability to create an Embed from a dict
- Added missing type hints
- Added API documentation for `from_dict()` and `asdict()`

### Changes

- Removed support for Python 3.4
- Renamed `to_dict()` to `asdict()`

## [0.4.1] - 2020-06-15

### Fixed

- Wrong PyPI upload format

## [0.4.0] - 2020-06-14

### Added

- Automatic retries on retry-able HTTP errors with exponential backoff
- Timeout for requests
- Included response headers in debug log

## [0.3.2] - 2020-03-31

### Fixed

- Removed deprecated attributes, added missing test cases

## [0.3.1] - 2019-12-07

### Fixed

- pip install error when "requests" module was not already present

## [0.3.0] - 2019-12-04

### Added

- Webhooks returns HTTP status code and headers. This allows the implementation of rate limit handling.

### Changed

- Webhook.execute() now returns a response object instead of the send report. The response objects contains the HTTP status code, headers and content of the response from Discord (e.g. send report)

- Webhook.execute() wil no longer throw an exception on non 2xx HTTP codes. Instead the application code needs to investigate the status code that is returned. It will still through exceptions on all network related issues.

## [0.2.0] - 2019-11-12

### Added

- Initial


examples/examples.py
import datetime
import logging
import os

from dhooks_lite import Webhook, Embed, Footer, Image, Thumbnail, Author, Field


logging.basicConfig(format="%(levelname)s: %(message)s", level=logging.INFO)

if "DISCORD_WEBHOOK_URL" not in os.environ:
    raise ValueError('mandatory environment variable "DISCORD_WEBHOOK_URL" not set')
else:
    DISCORD_WEBHOOK_URL = os.environ["DISCORD_WEBHOOK_URL"]


# Minimal example: Hello World
hook = Webhook(DISCORD_WEBHOOK_URL)
hook.execute("Hello, World!")

# Set username and avatar URL
hook = Webhook(
    DISCORD_WEBHOOK_URL,
    username="Bruce Wayne",
    avatar_url="https://i.imgur.com/thK8erv.png",
)
hook.execute("I am Batman!")

# Minmal embed example
hook = Webhook(DISCORD_WEBHOOK_URL)
e = Embed(description="Simple Embed example")
hook.execute(embeds=[e])


# Example with two embeds and all parameters
hook = Webhook(DISCORD_WEBHOOK_URL)
e1 = Embed(
    description="Only a few years ago, scientists stumbled upon an electrical current of cosmic proportions: 10^18 amps, or roughly one trillion lightning bolts. The lightning is thought to originate from an enormous black hole in the center of the galaxy, which has a core that is supposedly a “huge cosmic jet.” Apparently, the black hole’s huge magnetic field allows it to fire up this lightning bolt through gas and dust to a distance of over one hundred and fifty thousand light years away. And we thought that our galaxy was big, this single lightning bolt is one and a half times the size of it.",  # noqa
    title="Universe's highest electric current found",
    url="https://www.newscientist.com/article/mg21028174-900-universes-highest-electric-current-found/",  # noqa
    timestamp=datetime.datetime.utcnow(),
    color=0x5CDBF0,
    footer=Footer("Science Department", "https://i.imgur.com/Bgsv04h.png"),
    image=Image("https://i.imgur.com/eis1Y0P.jpg"),
    thumbnail=Thumbnail("https://i.imgur.com/2A4k28x.jpg"),
    author=Author(
        "John Scientist",
        url="https://en.wikipedia.org/wiki/Albert_Einstein",
        icon_url="https://i.imgur.com/1JoHDw1.png",
    ),
    fields=[Field("1st Measurement", "Failed"), Field("2nd Measurement", "Succeeded")],
)
e2 = Embed(description="TOP SECRET - Do not distribute!")

response = hook.execute(
    "Checkout this new report from the science department:",
    username="Bruce Wayne",
    avatar_url="https://i.imgur.com/thK8erv.png",
    embeds=[e1, e2],
    wait_for_response=True,
)
print(response.headers)
print()
print(response.content)
